azure/src/main/java/org/apache/iceberg/azure/adlsv2/AzureSasCredentialRefresher.java

```diff
@@ -0,0 +1,69 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.azure.adlsv2;
+
+import com.azure.core.credential.AzureSasCredential;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class AzureSasCredentialRefresher {
+  private final Supplier<Pair<String, Long>> sasTokenWithExpirationSupplier;
+  private final ScheduledExecutorService refreshExecutor;
+
+  private final AzureSasCredential azureSasCredential;
+
+  private static final long MAX_REFRESH_WINDOW_MILLIS = 300_000; // 5 minutes;
+  private static final long MIN_REFRESH_WAIT_MILLIS = 10;
+
+  public AzureSasCredentialRefresher(
+      Supplier<Pair<String, Long>> sasTokenWithExpirationSupplier,
+      ScheduledExecutorService refreshExecutor) {
+    this.sasTokenWithExpirationSupplier = sasTokenWithExpirationSupplier;
+    this.refreshExecutor = refreshExecutor;
+    Pair<String, Long> sasTokenWithExpiration = sasTokenWithExpirationSupplier.get();
+    this.azureSasCredential = new AzureSasCredential(sasTokenWithExpiration.getLeft());
+    scheduleRefresh(System.currentTimeMillis(), sasTokenWithExpiration.getRight());
+  }
+
+  public AzureSasCredential get() {
+    return this.azureSasCredential;
+  }
+
+  private void scheduleRefresh(long startTimeMillis, Long expireAtMillis) {
+
+    long expireInMillis = expireAtMillis - startTimeMillis;
+    long refreshWindowMillis = Math.min(expireInMillis / 10, MAX_REFRESH_WINDOW_MILLIS);
+    long waitIntervalMillis = expireInMillis - refreshWindowMillis;
+    long elapsedMillis = System.currentTimeMillis() - startTimeMillis;
+    long timeToWait = Math.max(waitIntervalMillis - elapsedMillis, MIN_REFRESH_WAIT_MILLIS);
+    this.refreshExecutor.schedule(
+        () -> {
+          long refreshStartTime = System.currentTimeMillis();
+          Pair<String, Long> sasTokenWithExpiration = sasTokenWithExpirationSupplier.get();
+          azureSasCredential.update(sasTokenWithExpiration.getLeft());
+          if (sasTokenWithExpiration.getRight() != null) {
+            this.scheduleRefresh(refreshStartTime, sasTokenWithExpiration.getRight());
+          }
+        },
+        timeToWait,
+        TimeUnit.MILLISECONDS);
```

Comment:
I follow the logic but it feels a bit complex, and I'm not sure about recursively scheduling in this manner...I feel like having a separate `refreshDelay()` or something just for getting the `timeToWait` would make this easier to read. I'll try and implement a concrete example

Reply:
Thanks for the suggestion, extracted a new refreshDelay() method to compute timeToWait.

> I'm not sure about recursively scheduling in this manner

I am referring to similar logic implemented in scheduleTokenRefresh method in [OAuth2Util](https://github.com/apache/iceberg/blob/apache-iceberg-1.7.0/core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java#L613-L640)